src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,344 @@
+
+  private final String       workerId;
+  private final SecureRandom random;
+
+  private final UnlockOperation unlockOperation;
+
+  private boolean running;
+  private boolean finished;
+
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
+                             ReplicatedJedisPool jedisPool,
+                             AccountsManager accountsManager)
+      throws IOException
+  {
+    DirectoryConfiguration directoryConfig = cdsConfig.getDirectoryConfiguration();
+    this.serverApiUrl    = directoryConfig.getServerApiUrl();
+    this.jedisPool       = jedisPool;
+    this.accountsManager = accountsManager;
+
+    this.random = new SecureRandom();
+
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    this.workerId = Hex.toString(workerIdBytes);
+
+    this.unlockOperation = new UnlockOperation(jedisPool);
+
+    SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                               .securityProtocol("TLSv1.2");
+    try {
+      sslConfig.trustStore(initializeKeyStore(directoryConfig.getServerApiCaCertificate()));
+    } catch (CertificateException ex) {
+      logger.error("error reading serverApiCaCertificate from contactDiscovery config", ex);
+    }
+
+    this.client = ClientBuilder.newBuilder()
+                               .register(HttpAuthenticationFeature.basic("signal", directoryConfig.getServerApiToken().getBytes()))
+                               .sslContext(sslConfig.createSSLContext())
+                               .build();
+  }
+
+  private static KeyStore initializeKeyStore(String pemCaCertificate)
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public synchronized void start() throws Exception {
+    running  = true;
+    finished = false;
+    new Thread(this).start();
+  }
+
+  @Override
+  public synchronized void stop() throws Exception {
+    running = false;
+    while (!finished) Util.wait(this);
+  }
+
+  private synchronized boolean sleepWhileRunning(long delay) {
+    long start   = System.currentTimeMillis();
+    long elapsed = 0;
+    while (running && elapsed < delay) {
+      try {
+        wait(delay - elapsed);
+      } catch (InterruptedException ex) {
```

Comment:
We never use interrupts, so you can assert here or just Util.wait()

Reply:
done in df3715170b6d1fd6c7abbfe340c18d8c1c0336be